DDD-7.md

```diff
@@ -0,0 +1,153 @@
+
+* allow to run multiple [source tasks](https://github.com/apache/kafka/blob/3.6.0/connect/api/src/main/java/org/apache/kafka/connect/source/SourceTask.java) for given connector if the connector provides multiple tasks
+* run potentially time-consuming code (e.g. event transformation or serialization) in the dedicated threads
+* allow possible further speedup by optionally disabling total ordering of the messages
+* be well-prepared for future changes and new features, namely
+  * switching to [virtual threads](https://docs.oracle.com/en/java/javase/21/core/virtual-threads.html)
+  * eventually delegate source tasks to external workers via [gRPC](https://grpc.io/) calls (especially to be able to horizontally scale connector load to multiple pods in Kubernetes cluster)
+  * better integration with Debezium k8s operator and UI
+  * providing Debezium engine as a [Quarkus](https://quarkus.io/) extension
+
+Another high-level goal is to adjust the current Debezium testsuite to use the `DebeziumEngine` interface instead of hard-coded `EmebddedEngine` implementation.
+This should allow for easy switching to any other implementation of `DebeziumEngine` and thus making new implementation to be easily tested with the current testsuite.
+
+## Non-goals
+
+* Change [`DebeziumEngine`](https://github.com/debezium/debezium/blob/v2.4.0.Final/debezium-api/src/main/java/io/debezium/engine/DebeziumEngine.java) interface.
+* Do any other changes in the Debezium [API package](https://github.com/debezium/debezium/tree/v2.4.0.Final/debezium-api).
+* Implement any parallelization inside connectors (e.g. tracking database CDC changes in multiple threads).
+* Remove dependency on Kafka Connect API.
+
+### Preserving Kafka Connect model
+
+As the main goal of the Debezium engine is to be able to execute Debezium outside of Kafka, it may be strange why not to take this opportunity to get rid of Kafka dependencies.
+The reason is simple: it would be a too complex change which would not impact only the Debezium engine.
+E.g. removing [`WorkerConfig`](https://github.com/apache/kafka/blob/3.6.0/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConfig.java) would require removing [`OffsetBackingStore`](https://github.com/apache/kafka/blob/3.6.0/connect/runtime/src/main/java/org/apache/kafka/connect/storage/OffsetBackingStore.java),
+which would require removing [`OffsetStorageReader`](https://github.com/apache/kafka/blob/3.6.0/connect/api/src/main/java/org/apache/kafka/connect/storage/OffsetStorageReader.java) etc., etc., resulting in substantial changes in the Debezium core and connectors.
+Therefore, this should be done in a separate task which would deserve a dedicated DDD describing all the changes and proposing replacements for Kafka Connect interfaces and classes.
+
```

Comment:
I've already started with implementation and one thing which complicates the code little bit is that Connect model doesn't allow access to connector and task contexts. Would it make sense to replace it with our interfaces and implementation as a preparation step to abandon Connect model?

It would look like this:
```
@Incubating
public interface DebeziumSourceConnectorContext {

    /**
     * Returns the {@link OffsetStorageReader} for this DebeziumConnectorContext.
     * @return the OffsetStorageReader for this connector.
     */
    OffsetStorageReader offsetStorageReader();

    /**
     * Returns the {@link OffsetStorageWriter} for this DebeziumConnectorContext.
     * @return the OffsetStorageWriter for this connector.
     */
    OffsetStorageWriter offsetStorageWriter();

    /**
     * Notifies the engine it should rise an error and allows engine to take appropriate action.
     * @param e Exception to be raised.
     */
    void onError(Exception e);
}
```
and implementation implementing both our and Connect interfaces:
```
public static final class EngineSourceConnectorContext implements DebeziumSourceConnectorContext, SourceConnectorContext
```
so that we have our own interface and at the same time we can still pass it to the other parts which require Connect model.

It's harder for part which are not interfaces, e.g. `SourceTask`, in such case we can have a delegate instance inside it:
```
public static final class EngineSourceConnector implements DebeziumSourceConnector {

        private final SourceConnector connectConnector;
        private DebeziumSourceConnectorContext context;

        public EngineSourceConnector(SourceConnector connectConnector) {
            this.connectConnector = connectConnector;
        }

        public SourceConnector connectConnector() {
            return connectConnector;
        }

        @Override
        public DebeziumSourceConnectorContext context() {
            return this.context;
        }

        @Override
        public void initialize(DebeziumSourceConnectorContext context) {
            this.context = context;
            this.connectConnector.initialize((SourceConnectorContext) context);
        }
    }
```
In a similar way we can probably handle storage part for now (as a preparation step - use our interface, implementation having also a Connect delegate and pass Connect delegate where needed).

Does it make sense to you @jpechane ? If yes, should I add it here or should be another DDD?

Reply:
Let's add it here